deps!: BigQuery Storage and pyarrow are required dependencies

plamut opened the pull request:
Closes #757.

This might already be a final version, but I still need to check a few additional things:
 - ~Remove logic for "ensure client"? Or maybe not, since we still allow tabledata.list?~
   We still need that.
 - ~Remove min version check for BQ Storage, since now it is guaranteed `>=2.0.0` will be installed (no more `pip` quirks with extras).~
   Done.
 - ~Remove checks for GAPIC? It's now a required dependency.~
   Not applicable, the minimum GAPIC version was never checked at runtime.

**PR checklist:**:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

plamut committed:
process: make BQ Storage and pyarrow required

plamut committed:
Make pyarrow required in _pandas_helpers.py

plamut committed:
Make pyarrow required in client.py

plamut committed:
Make pyarrow required in table.py

plamut committed:
Make pyarrow required in job/query.py

plamut committed:
Make pyarrow required in DBAPI tests

plamut committed:
Make pyarrow required in snippets tests

plamut committed:
Make BQ storage required in client.py

plamut committed:
Make BQ storage required in table.py

plamut committed:
Make BQ storage required in magics.py

plamut committed:
Make BQ storage required in test__helpers.py

plamut committed:
Make BQ storage required in test__pandas_helpers.py

plamut committed:
Make BQ storage required in test_query_pandas.py

plamut committed:
Make BQ storage required in DB API tests

plamut committed:
Make method signatures compatible again

The annotations caused a mismatch

plamut committed:
Remove checks for minimum BQ Storage version

Since this is now a required dependency, there should not be any more
pip quirks that used to allow installing BQ Storage as an extra, but
without always respecting its minimum version pin.

plamut commented:
@tswast It might be easier to review each commit separately, I tried to make them independent from each other. Also, double check if the assumption stated in the last commit's message really holds - if not, we can revert that bit.

tseaver commented on setup.py:
```diff
@@ -32,24 +32,18 @@
     "google-resumable-media >= 0.6.0, < 3.0dev",
     "packaging >= 14.3",
     "protobuf >= 3.12.0",
+    "pyarrow >= 1.0.0, < 5.0dev",
```
Making these previously-optional dependencies now required feels like it might be a `semver: major` change.

plamut replied:
IIRC @tswast said in a chat that it shouldn't be a problem for most users, and those who cannot install these dependencies for any reason can still pin to a lower (non-major version). But reading this as I write, we might actually want to re-consider this.

plamut replied:
There has been a [debate](https://github.com/googleapis/python-bigquery/issues/757#issuecomment-879908013) on this on the issue itself. Also copying Tim's input:

> The gist is that the problematic dependencies are already required transitively anyway
With pyarrow being the exception
It's probably worth the bump for pyarrow if we're being conservative about it. Esp since we do know of environments where adding it will break

tswast replied:
Let's go with the major version bump. I'd rather people adopt this change a little more slowly than break someone who has a working 2.x branch without pyarrow

plamut replied:
Roger.

Is there perhaps anything else that we want to clean up and would be a breaking change? (possibly in a separate PR)
A major version bump is a good opportunity for that.

tswast replied:
https://github.com/googleapis/python-bigquery/issues/319 is still on my mind. Perhaps easier now that proto-plus added more case options?

tswast replied:
Another breaking change we might want to do: https://issuetracker.google.com/144712110 -- Better type conversions in the pandas connector.

plamut replied:
And yet another one from today - maybe bumping the minimum `pyarrow` version, as the older ones suffer from a [bug](https://github.com/googleapis/python-bigquery/issues/781#issuecomment-883497414) (load data from dataframe can upload incorrect data, yikes!)

tswast replied:
That is a nasty bug. Thankfully `pyarrow` didn't follow [NEP-29](https://numpy.org/neps/nep-0029-deprecation_policy.html) this time and kept support for Python 3.6 in the 4.x packages. I'd be okay with bumping the minimum for that.

plamut replied:
And yet another - if this gets delayed significantly (read: months), we might want to consider dropping Python 3.6 support as well. But Python 3.6 [end of life](https://www.python.org/dev/peps/pep-0494/#id15) will only happen in December this year.

plamut commented on google/cloud/bigquery/exceptions.py:
```diff
@@ -18,4 +18,8 @@ class BigQueryError(Exception):
-    """Raised when too old a version of BigQuery Storage extra is detected at runtime."""
+    """Raised when too old a version of BigQuery Storage extra is detected at runtime.
+
+    .. deprecated:: 2.2.0
+       Not used anymore and will be removed in the future.
```
Note to self - this should have been 2.22.0, but it's not obsolete, since we've just made that [release](https://github.com/googleapis/python-bigquery/pull/771#issuecomment-882856588). Needs to be adjusted once we decide what the next version will be.

plamut replied:
Done. Removed the custom exception class altogether, because we'll do a major version bump, thus a breaking change is acceptable.

plamut commented on setup.py:
```diff
@@ -32,24 +32,18 @@
     "grpcio >= 1.38.1, < 2.0dev",  # https://github.com/googleapis/python-bigquery/issues/695
     "google-api-core[grpc] >= 1.29.0, < 3.0.0dev",
     "proto-plus >= 1.10.0",
+    "google-cloud-bigquery-storage >= 2.0.0, <3.0.0dev",
```
This should still allow 1.x, as per https://github.com/googleapis/python-bigquery/pull/784/files

It also means we probably need to add the minimum version check back. Or wait for https://github.com/googleapis/google-cloud-python/issues/10566 to be resolved.

plamut committed:
Merge branch 'master' into iss-757

plamut committed:
Remove LegacyBigQueryStorageError

Since it will be released in a major version bump, we can make this
a breaking change, i.e. without deprecation.

busunkim96 commented:
Hi,

Stopping by to follow up on https://github.com/googleapis/google-cloud-python/issues/10566#issuecomment-883215649. That issue is aimed specifically at `google-cloud-core`, `google-api-core`, and `google-auth` libraries. Are you anticipating having to require >=2.x.x for any of those libraries to support this change?

plamut commented:
@busunkim96 I edited that [comment](https://github.com/googleapis/python-bigquery/pull/776#discussion_r672860955). :)

I don't think so, we are just making a few extras required now, nothing to do with core dependencies (wrote that comment too fast).

tswast commented on tests/unit/test__pandas_helpers.py:
```diff
@@ -29,13 +29,9 @@
     import pandas.testing
 except ImportError:  # pragma: NO COVER
     pandas = None
-try:
```
I notice a test failure:

```
________________ test_dataframe_to_parquet_dict_sequence_schema ________________

module_under_test = <module 'google.cloud.bigquery._pandas_helpers' from '/tmpfs/src/github/python-bigquery/google/cloud/bigquery/_pandas_helpers.py'>

    @pytest.mark.skipif(isinstance(pyarrow, mock.Mock), reason="Requires `pyarrow`")
    def test_dataframe_to_parquet_dict_sequence_schema(module_under_test):
        dict_schema = [
            {"name": "field01", "type": "STRING", "mode": "REQUIRED"},
            {"name": "field02", "type": "BOOL", "mode": "NULLABLE"},
        ]

>       dataframe = pandas.DataFrame(
            {"field01": [u"hello", u"world"], "field02": [True, False]}
        )
E       AttributeError: 'NoneType' object has no attribute 'DataFrame'

tests/unit/test__pandas_helpers.py:1252: AttributeError
```

Can we use `importorskip` here for pandas?

plamut replied:
Done.

tswast commented on tests/unit/test__pandas_helpers.py:
```diff
@@ -1093,33 +1077,6 @@ def test_dataframe_to_parquet_compression_method(module_under_test):
-
-    # a warning should also be issued
-    expected_warnings = [
-        warning for warning in warned if "could not determine" in str(warning).lower()
```
Looks like the warning for this isn't issued anymore. I wonder if it's possible to get a dataframe without detectable schema if pyarrow is installed?

plamut replied:
This particular test covered the [case](https://github.com/googleapis/python-bigquery/blob/e58702967d572e83b4c774278818302594a511b7/google/cloud/bigquery/_pandas_helpers.py#L346-L352) when we couldn't even start detecting the schema with `pyarrow` because the later is missing. This is not possible anymore, as `pyarrow` is now required, thus that particular warning has been removed.

However, we still call [augment_schema()](https://github.com/googleapis/python-bigquery/blob/e58702967d572e83b4c774278818302594a511b7/google/cloud/bigquery/_pandas_helpers.py#L354-L411) immediately after the deleted lines. That function can also issue a warning on its own if schema detection fails for some fields, but those paths are covered by other tests.

tswast commented:
I'm pretty sure it would be possible to delete even more code such as https://github.com/googleapis/python-bigquery/blob/d1cbc3817a4b93f61356bd14ba51fb176e5d0269/google/cloud/bigquery/_pandas_helpers.py#L542-L554 if we always have a BigQuery Storage API client, but there are a few cases where it can't be used.

For example: `list_rows(table).to_dataframe()` won't be able to use BQ Storage API from free tier. Also, when we implement https://github.com/googleapis/python-bigquery/issues/589 we'll need to use the REST API for that (at least in the single page of results case)

plamut committed:
Bump minimum pyarrow version to 3.0.0

plamut committed:
Remove unneeded pytest.importorskip for BQ Storage

plamut commented:
As a follow-up to https://github.com/googleapis/python-bigquery/issues/781 I also bumped the minimum `pyarrow` version.

tseaver commented:
Failing systest is another instance of #733.

plamut committed:
Merge branch 'master' into iss-757

plamut committed:
Remove pyarrow version checks in pandas helpers tests

plamut committed:
Conditionally skip pandas tests where needed

plamut committed:
Remove unneeded conditional pyarrow version paths

plamut committed:
Cover schema autodetect failed code path in test

plamut commented on google/cloud/bigquery/client.py:
```diff
@@ -2576,16 +2572,6 @@ def load_table_from_dataframe(
-                        "version or a different source format. "
-                        "See: https://github.com/googleapis/python-bigquery/issues/781"
-                    )
-                    warnings.warn(msg, category=RuntimeWarning)
```
This warning was removed because the just-fixed bug will not come into play once we require `pyarrow > 3.0.0`.

tswast committed:
Merge branch 'master' into iss-757

tswast committed:
fix bad merge

tswast reviewed (approved):
This is awesome!

FYI: I pointed this PR at the `v3` branch. Should be safe to merge.
